src/metabase/driver/postgres.clj

```diff
@@ -352,11 +353,26 @@
     (pretty [_]
       (format "%s::%s" (pr-str expr) (name psql-type)))))
 
-(defn- json-query [identifier nfc-path]
+(defn- json-query [identifier nfc-field]
   (letfn [(handle-name [x] (if (number? x) (str x) (name x)))]
-    (apply hsql/call [:json_extract_path_text
-                      (hx/cast :json (keyword (first nfc-path)))
-                      (mapv #(hx/cast :text (handle-name %)) (rest nfc-path))])))
+    (let [field-type           (:effective_type nfc-field)
+          nfc-path             (:nfc_path nfc-field)
+          unwrapped-identifier (:form identifier)
+          parent-identifier    (field/nfc-field->parent-identifier unwrapped-identifier nfc-field)
+          ;; Array and sub-JSON coerced to text
+          cast-type            (cond
+                                 (isa? field-type :type/Integer)
+                                 :type/Integer
+                                 (isa? field-type :type/Float)
+                                 :type/Float
+                                 (isa? field-type :type/Boolean)
+                                 :type/Boolean
```

Comment:
wondering if there's a specific reason to keep the `:type` namespace on the keywords used for `cast-type`? seems like `hx/cast` calls `name` on them anyway which discards the namespace. it's functionally the same so don't need to change it, more just curious

Reply:
not really. I've been thinking of them in namespace so I keep them around in case I need that but I probably won't